core/utils/utils.go

```diff
@@ -418,46 +418,48 @@ func WaitGroupChan(wg *sync.WaitGroup) <-chan struct{} {
 // channel is closed.
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
-func WithCloseChan(parentCtx context.Context, chStop <-chan struct{}) (ctx context.Context, cancel context.CancelFunc) {
-	ctx, cancel = context.WithCancel(parentCtx)
-
-	go func() {
-		select {
-		case <-chStop:
-		case <-ctx.Done():
-		}
-		cancel()
-	}()
-
-	return ctx, cancel
+func WithCloseChan(parentCtx context.Context, chStop <-chan struct{}) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).CancelCtx(context.WithCancel(parentCtx))
 }
 
 // ContextFromChan creates a context that finishes when the provided channel
 // receives or is closed.
-// When channel closes, the ctx.Err() will always be context.Canceled
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
 func ContextFromChan(chStop <-chan struct{}) (context.Context, context.CancelFunc) {
-	ctx, cancel := context.WithCancel(context.Background())
-	go func() {
-		select {
-		case <-chStop:
-			cancel()
-		case <-ctx.Done():
-		}
-	}()
-	return ctx, cancel
+	return StopChan(chStop).NewCtx()
 }
 
-// ContextFromChanWithDeadline creates a context with a deadline that finishes when the provided channel
+// ContextFromChanWithTimeout creates a context with a timeout that finishes when the provided channel
 // receives or is closed.
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
-func ContextFromChanWithDeadline(chStop <-chan struct{}, timeout time.Duration) (context.Context, context.CancelFunc) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+func ContextFromChanWithTimeout(chStop <-chan struct{}, timeout time.Duration) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).CancelCtx(context.WithTimeout(context.Background(), timeout))
+}
+
+type StopChan <-chan struct{}
+
+func NewStopChan() (ch StopChan, stop func()) {
+	c := make(chan struct{})
+	return c, func() { close(c) }
+}
+
+// NewCtx returns a new background [context.Context] that will be cancelled when the StopChan is closed.
+//
+// Note: The [context.CancelFunc] must be called to prevent memory leaks.
+func (s StopChan) NewCtx() (context.Context, context.CancelFunc) {
+	return s.CancelCtx(context.WithCancel(context.Background()))
+}
+
+// CancelCtx spawns a goroutine to cancel a [context.Context] when StopChan is closed.
```

Comment:
same. a StopChan can't be `closed` directly by a caller, instead stop fn must be called

Reply:
The owner of `StopChan` can still hold a direct reference to the original `chan struct{}`.

Reply:
isn't that confusing to a dev? multiple way to do have the same runtime effect using completely different types and functions?

instead of type casting and leading to ambiguous usage patterns, i'd prefer taking ownership of any underlying channel

Reply:
If something is confusing than we can adjust the API and docs, but imposing one usage pattern necessarily leaves out use cases that are meant to be covered here. This is flexible by design because we have existing code that I didn't want to make noisier. The constructor is a special case for convenience.